[TEAM #08 FE] 3주차 PR

Kakamotobi opened the pull request:
# 👟 Issue Tracker 3주차
[8조 이슈 트래커 배포 링크 바로가기](http://issue-tracker-8.s3-website.ap-northeast-2.amazonaws.com/)

## 실행 방법
```zsh
git clone https://github.com/issue-tracker-08/issue-tracker-max.git
git switch fe-w3
cd fe
npm i
npm run dev
```

## 3주차 Milestone
- 이슈 상세 페이지
- 레이블 목록 페이지
- 마일스톤 목록 페이지

## 관련 Issues
- [#125 [FE] 마일스톤 페이지 UI & 기능](https://github.com/issue-tracker-08/issue-tracker-max/issues/125)
- [#104 [FE] LabelPage](https://github.com/issue-tracker-08/issue-tracker-max/issues/104)
- [#106 [FE] 이슈 내용 조회, 수정 & 댓글 조회, 수정](https://github.com/issue-tracker-08/issue-tracker-max/issues/106)

## IssuePage 상태관리
아래 방식 중 현재 첫번째 방식으로 하고 있지만 두번째 방식으로 바꿀 예정.
1) Issue 수정/삭제(POST/DELETE)가 성공하면 `navigate(0)`으로 reload (다시 fetch). (현재 방식)
2) `useFetch`에서 `refetch`를 제공하여 성공하면 invoke. (개선할 방식)
3) 성공하면 성공 값으로 IssuePage에 있는 `issuesList` 상태를 직접 update.

## 로그인 User 정보
- 현재 상황
  - 로그인 성공시 서버로부터 access token과 user 정보 받음.
- 개선할 방향
  - 로그인 성공시 서버로부터 access token 받음.
  - 성공하자마자 서버로 user 정보 요청 (access token 동반).

## Null checking
- Data가 아직 안받아져왔을 때 등, data가 null인 경우가 있는데, null-checking의 적절한 위치.
- Loading 상태를 안쓰고 있는데 data를 사용하는 컴포넌트들에서 null을 어떻게 처리할지 고민.

youzysu committed:
[FE] useFetch hook 개선, IssueDetailBody 변경 사항 반영 (#99)

* #94 refactor: useFetch 수정

- `isLoading`, `error` 추가

* #94 refactor: useFetch data initial type 수정

- nullable X

* #94 refactor: useFetch 사용처(IssuePage, 컴포넌트) 변경 사항 반영

* #94 refactor: 이슈 디테일 상태 관리 접근 조정

- `IssueDetailPage`단에서 상세 detail 정보 fetch.
  - `/details`, `/sidebar`, `/comments`

* #99 refactor: IssueDetailPage type, mock data, api 변경사항 반영

---------

Co-authored-by: Daeram Chung <daeram.chung@gmail.com>

Kakamotobi committed:
[FE] IssueDetailHeader 컴포넌트 구현 (#101)

* #79 feat: Issue 제목, 열림/닫힘 API 추가

* #79 feat: Issue 제목, 열림/닫힘 수정 mock 추가

* #79 fix: TextInput의 input에 value 전달

* #79 feat: IssueDetailHeader 컴포넌트 구현

- 부모인 `IssueDetailPage`로부터 `issueDetails` 객체 및 수정 함수들을 props로 전달 받음.

* #79 refactor: issueDetails에 대한 null check을 IssueDetailHeader로 이동

- 기존: `issueDetails && <IssueDetailHeader />`
- 수정: `IssueDetailHeader` 내부에서 `issueDetails`가 null 일시에 초기 값 설정.

* #79 design: Label width 설정

* #79 fix: issueDetails 업데이트/fetch 완료 시 newTitle 상태 업데이트

- 공식문서의 권고에 따라, `useEffect`을 사용하지 않고 이전 prop을 상태로 두고 과 현재 prop과 비교하여 다르면 `newTitle` 초기 값 업데이트.

Kakamotobi committed:
[FE] Label Page 중간 PR (#116)

* #104 design: TableHeaderLabel 컴포넌트 구현

- `TableHeader` 템플릿 컴포넌트를 사용하여 `TableHeaderLabel` 컴포넌트 구현.

* #104 design: TableBodyItemLabel 컴포넌트 구현

- `TableBodyItem` 컴포넌트를 사용하여 `TableBodyItemLabel` 컴포넌트 구현.

* #104 design: danger filter 추가

* #104 design: 빈 table body item 구현

* #104 design: TableBodyLabels 구현

* #104 design: LabelPage 구현

* #104 rename: Table 관련 모듈 파일 정리

youzysu committed:
[FE] 이슈 상세 페이지 상태 관리 & 댓글 기능 (#112)

* #106 refactor: IssueDetailPage 상태 관리 방식 변경 사항 적용

* #106 feat: 이슈 상세 페이지 Content Comment 렌더링

- ReactMarkdown 추가

* #106 feat: 댓글 목록 조회 기능

* style: 사이드바 min-width, max-height 설정

* style: TextArea, Label styled component style 수정

- Label.style.ts로 변경, width 고정값 제거
- input flex-grow, textArea width 100% 추가

* #106 feat: 댓글 작성 및 추가

* #106 refactor: TextArea 하위 컴포넌트 분리

- Comment 수정 시 하위 컴포넌트만 재사용하기 위해 분리

* #106 fix: 이후 댓글 있는 경우에만 추가 요청하도록 수정

* #106 feat: 이슈 내용/댓글 수정 기능

- 현재 로그인 유저와 댓글 작성자가 동일할 때만 편집 버튼 보여주기
- 이슈 상세 페이지 Body 하위 컴포넌트 분리: Comment 영역 / Sidebar 영역
- 함수명 수정

* #112 fix: 댓글 추가 응답 내용 반영

- 저장된 userInfo 사용

* #106 style, refactor:

- DropdownPanel max-height 설정
- Issue Assignees, Labels 수정사항 없을 때 수정 요청 보내지 않도록 수정

youzysu committed:
feat: NotFoundPage 추가 (#124)

youzysu committed:
#106 feat: 이슈 댓글 개수 추가 (#128)

Kakamotobi committed:
[FE] Label Page 및 Label Editor (#131)

* #104 fix: label 조회 mock에 description 추가

* #104 feat: hex color 관련 util 함수 구현

- `isValidHexColor`: 사용자가 직접입력한 값이 유효한 hex color인지 확인.
- `generateRandomHexColor`: 랜덤한 hex color 생성.

* #104 design: display bold 20 font 추가

* #104 refactor: LabelTag props 수정

* #104 feat: Label 생성, 수정, 삭제 API 요청 함수 추가

* #104 fix: TextInput 스타일 및 props 수정

- `TextInput` 내부의 실제 input에 전해줄 attributes는 제한해두고, 나머지 html attributes는 최상단 `StyledTextInput`가 받는 것으로 변경.
- `TextInput`에서 children 받기.
- `Label` 최소 width 추가.

* #104 feat: labels 생성, 수정, 삭제 mock handler 추가

* #104 feat: hex color가 밝은지 판별하는 함수

* #104 design: LabelTag 수정된 props 반영

* #104 design: Dropdown UI 수정

* #104 rename: decToHex -> decToHexOpacity

* #104 design: Fieldset width 제거

* #104 feat: LabelEditor 컴포넌트 구현

* #104 fix: deleteLabel method 수정

* #104 feat: LabelEditor 생성 및 수정시 적용

youzysu committed:
[FE] Table 컴포넌트 구조 개선 & MilestoneTable (#130)

* #125 feat, style: Table 컴포넌트 구조 개선 & MilestoneTable

- Table 컴포넌트 Issue, Label, Milestone 단위로 디렉토리 구조 변경
- 마일스톤 목록 전체 조회 UI

* #125 rename: Table Component names 수정

* fix: TextInput type props 추가

Kakamotobi committed:
fix: LabelEditor input trim bug 고침

Kakamotobi committed:
fix: Label description optional로 수정

Kakamotobi committed:
fix: useFetch에서 statusText가 아닌 status로 fetch 성공 여부

youzysu committed:
[FE] API 연동 관련 수정 (#135)

* fix: 로그인 상태 유지

* fix: Sidebar API 필드명 백엔드와 통일

- suffix Ids

* fix: DropdownItem key content -> id로 수정

* fix: 요청 성공 시 statusText -> status로 수정

* fix: 메인화면 이슈 선택 시 이슈 상세 페이지로 이동 Link 태그로 변경

- a 태그 href 시 서버로 해당 리소스 요청됨

* fix: 이슈 상세 조회 API path 수정

- details 삭제

youzysu committed:
fix: LabelTag key LabelId -> name

- Label Type 수정 필요
- 메인 화면 labels는 Label Type 아님

youzysu committed:
feat: 로그아웃 버튼 추가, 회원가입 후 라우팅 수정

Kakamotobi committed:
[FE] 배포 hot fix (#140)

* fix: patchLabel -> putLabel

* design: 로그아웃 버튼 텍스트 색상 적용

* #design: 전체 틀 반응형 적용

* fix: IssueItem author type 수정

* IssueDetailHeader 제목 수정 TextInput name 수정

youzysu committed:
[FE] 마일스톤 추가, 수정, 삭제 기능 및 기타 스타일 수정 (#136)

* #133 refactor: 로그아웃 버튼, 로그인 라우팅 수정, textarea 줄바꿈 css 속성 추가

* #125 feat: 마일스톤 추가 기능

* #125 feat: 마일스톤 수정 & 삭제 기능

* #133 refactor: 타입, 스타일 수정

- 마일스톤 없을 때도 빈 아이템 보여주도록 수정
- openIssueCount, closedIssueCount number 타입으로 수정

* refactor: 리뷰 반영

Kakamotobi committed:
[FE] 배포 hot fix 2 (#141)

* fix: patchLabel -> putLabel

* design: 로그아웃 버튼 텍스트 색상 적용

* #design: 전체 틀 반응형 적용

* fix: IssueItem author type 수정

* IssueDetailHeader 제목 수정 TextInput name 수정

* fix: SignupPage에서 LoginPage가는 링크 path 수정

* fix: AuthPage width 수정

* fix: Date 객체 생성하지 않고 Date.now()로 변경

youzysu commented on fe/src/components/Issues/IssueDetail/IssueCommentContainer.tsx:
```diff
@@ -0,0 +1,158 @@
+    useCallback(() => getComments(issueNumber, cursor), [issueNumber, cursor])
+  );
+
+  useEffect(() => {
+    issueComments && setAllComments((prev) => [...prev, ...issueComments.data]);
+  }, [issueComments]);
```
개발 환경인 StrictMode에서 useEffect가 두번 돌아서 Comments가 두번씩 추가되는 이슈가 있었어요. 일단 빌드 환경에서는 문제가 없는데, 현재 로직을 개선해야 할지 고민이 됩니다 🥲

hongbiii replied:
StrictMode의 특수성이니 어쩔 수 없긴 하겠는데 get요청을 다시 하는 방식으로 변경하면 해소가 될 것 같아요 !

youzysu replied:
여기서는 어차피 get 요청을 하고 있는 부분인데요, 저희가 comment에서 pagination을 적용해서 comment를 10개씩 요청하고 있고, 전체 comment를 allComments 상태로 관리하고 있어서요! 

현재 로직을 좀 더 설명드리면 cursor값이 변경되면 다시 10개의 comment를 요청하게 되고, allComents 배열에 accumulate하는데, strictmode에서 useEffect가 두번 돌면서 배열에 두번 추가돼요!

카카모토비가 피드백주시기로는 개발 환경에서도 문제가 없도록, 즉 useEffect가 두번 돌아도 문제 없도록 해야 한다고 하시는데 저는 아이디어가 안떠올라서.. 그럴 수 있는 방법이 있을까? 해서 여쭤봤습니다!

hongbiii replied:
요 문서가 조금 도움이 될까요??
https://react-ko.dev/learn/synchronizing-with-effects#fetching-data

youzysu commented on fe/src/components/Comment.tsx:
```diff
@@ -1,48 +1,159 @@
+    setNewContent(content);
+  };
+
+  const onEditComplete = async () => {
+    setIsEditing(false);
+    const body = { content: newContent };
+
+    try {
+      const { status } = commentId
+        ? await putIssueComment(issueNumber, commentId, body)
+        : await putIssueContent(issueNumber, body);
+
+      if (status === 200) {
+        updateContent(newContent, commentId);
+      }
+    } catch (error) {
+      // TODO: error handling
+      console.log(error);
+    }
+  };
```
사실상 Issue Content도 UI가 Comment와 동일해서 Comment 컴포넌트를 사용하고 있어요!
이슈의 댓글인 경우에만 optional로 commentId를 props로 받아서 commentId 여부에 따라 이슈 내용 수정, 이슈 댓글 수정 여부를 결정하게 되는데요,
이 부분을 어떻게 개선하면 좋을지 궁금해요.

개선하려는 이유는 현재 commentId가 undefined인 경우에도 이슈 내용 수정 요청이 가게 돼서 고민입니다..!

hongbiii replied:
commentId가 undefined 이다 === 이슈 내용(content)이다 === 이슈 내용 수정 요청이 가는 게 맞는 동작 아닌가욤 ??

youzysu replied:
오, 말씀하신 것처럼 commentId가 undefined 이다 === 이슈 내용(content)이다 라고 볼 수 있을까? 라는 의문이 든것인데요, 혹시라도 commentId에 오류가 생기는 경우(?)가 있을 수도 있지 않을까요? 그럴 때 의도하지 않게 content를 수정하는 요청이 가게 돼서요. 중복이 있더라도 함수를 분리하는게 최선일까요 흠

hongbiii replied:
commentId를 get 하는 과정에서 오류가 생기는 경우를 말씀하시는 거 맞나요??
그럴 때는 아예 에러 응답을 내려주면 괜찮지 않을까 싶은데..

함수 분리는 오류가 생기는 경우 때문이라기보다는 추후 코멘트와 컨텐트의 로직이 달라질 수 있기 때문에 말씀드린 거였어요!

youzysu commented on fe/src/pages/MainPage/IssueDetailPage.tsx:
```diff
@@ -1,68 +1,61 @@
-  const isFetchedComplete = !!issueObj.issueId;
+  const updateIssueTitle = (newTitle: string) => {
+    setIssueDetails((prev) => {
+      return prev ? { ...prev, title: newTitle } : prev;
+    });
+  };
 
-  // TODO: fetch comments
-  const comments: unknown[] = [];
+  const updateIssueIsOpen = () => {
+    setIssueDetails((prev) => {
+      return prev ? { ...prev, isOpen: !prev.isOpen } : prev;
+    });
+  };
 
-  const { assignees, labels, milestone } = issueObj;
+  const updateIssueContent = (newContent: string) => {
+    setIssueDetails((prev) => {
+      return prev ? { ...prev, content: newContent } : prev;
+    });
+  };
 
-  // TODO: 이슈 상세 조회 API ID만 받도록 개선 필요 (받은대로 그대로 props로 전달)
-  const assigneeIds = assignees.map((assignee) => assignee.userAccountId);
-  const labelIds = labels.map((label) => label.labelId);
-  const milestoneId = milestone.milestoneId;
+  const updateIssueCommentCount = () => {
+    setIssueDetails((prev) => {
+      return prev ? { ...prev, commentCount: prev.commentCount + 1 } : prev;
+    });
+  };
```
# 상태 관리 방식 고민

## 현재 방식
1. 프론트 상태 별도 관리
여기 이슈 상세 페이지처럼 요청 성공 시 직접 set

2. navigate(0)
라벨 페이지, 마일스톤 페이지
React router useNavigate로 현재 페이지 refresh

## 고려중인 개선 방향

3. useFetch 내부에서 reFetch 함수를 반환하여 다시 전체 조회 재요청해서 data update하기

hongbiii replied:
고려중인 개선 방향이 좋아보이네요! 👍

1은 요청 실패 시 화면에 보여지는 데이터와 실제 DB에 저장된 데이터가 달라질 위험이 있을 것 같고
2는 한번 깜빡일 것 같아요

hongbiii commented on fe/src/App.tsx:
```diff
@@ -7,9 +7,10 @@ import LabelPage from "@pages/MainPage/LabelPage";
+import NotFoundPage from "@pages/NotFoundPage";
 import GlobalStyle from "@styles/GlobalStyle";
 import { darkMode, lightMode } from "@styles/designSystem";
-import { ProtectedRoute } from "ProtectedRoute";
```
오? ProtectedRoute는 왜 삭제됐나요? 궁금!!

youzysu replied:
로그인 상태 유지하려고 시도하다가 일단 삭제하게 되었는데, Route 쪽은 전반적으로 개선해봐야 할 것 같아요!

hongbiii commented on fe/src/utils/time.ts:
```diff
@@ -45,3 +45,8 @@ export const convertPastTimestamp = (timestamp: string) => {
   }
 };
+
+export const validateDate = (dueDate: string) => {
+  const regex = /^(19|20)\d{2}-(0[1-9]|1[012])-(0[1-9]|[12][0-9]|3[01])$/;
+  return regex.test(dueDate);
+};
```
정규표현식 사용은 좋지만 한눈에 알아보기 힘든 경우가 많아요~
이럴 때는 주석을 추가해주거나 테스트코드를 추가해주면 해당 정규표현식을 이해하는 데에 도움이 됩니다!

hongbiii commented on fe/src/pages/NotFoundPage.tsx:
```diff
@@ -0,0 +1,42 @@
+import Button from "@components/common/Button";
+import styled from "styled-components";
+
+export default function NotFoundPage() {
```
404 페이지 좋습니다 👍👍👍

hongbiii commented on fe/src/hooks/useFetch.ts:
```diff
@@ -21,14 +23,16 @@ export default function useFetch<T>(
+        throw Error((res.data as AxiosError).message);
       } catch (error) {
-        alert(error);
+        setErrorMessage(error as string);
```
여기서 던진 에러가 catch에서 바로 잡히는 걸 의도하신 게 맞나용?

hongbiii commented on fe/src/mocks/data.ts:
```diff
@@ -165,50 +159,106 @@ export const labelList = [
+export const openMilestoneList = Array.from({ length: 10 }, (_, i) => {
+  return {
+    milestoneId: i + 1,
+    milestoneName: faker.lorem.words(),
```
faker 편하죵 크크

hongbiii commented on fe/src/pages/MainPage/IssueDetailPage.tsx:
지금 4개의 페이지가 모두 MainPage 아래에 있어서 좀 어색하게 느껴집니다!
메인은 말그대로 메인.. 하나여야 할 것 같은 느낌?
기획 상으로는 이슈 목록 페이지가 첫 화면이긴 하지만 보통은 진짜 홈화면이 있기 때문에
MainPage 디렉토리를 아예 삭제하고 4개의 페이지를 모두 pages/ 바로 아래에 두어도 괜찮을 것 같아요!

youzysu replied:
넵 Route랑 Page 관리는 전반적으로 리팩토링을 하려고 합니다!

hongbiii commented on fe/src/pages/MainPage/IssuesPage.tsx:
```diff
@@ -11,36 +13,38 @@ import { useNavigate } from "react-router-dom";
-  const numOpen = issuesList.filter((issue) => issue.isOpen).length;
-  const numClosed = issuesList.length - numOpen;
+  const { data: issuesList } = useFetch(getIssues);
+  const { data: labelsList } = useFetch(getLabels);
+  const { data: milestonesList } = useFetch(getMilestones);
```
[작은 의견]

개수를 구하기 위해 라벨, 마일스톤 목록을 전부 불러올 필요가 있을까 하는 생각이 듭니다!

youzysu replied:
저도 그렇게 생각합니다...😇 아래 말씀하신 것처럼 백엔드에 개수만 받아오는 API를 추가로 요청하거나, 혹은 프론트에서 관리할 수 있는 방법이 있을까요? 마일스톤 페이지, 레이블 페이지에서 모두 사용하고 있는 상태라서요.
떠오르는건 두 페이지의 상단인 MainPage에서 목록을 불러와서 context로 하위 페이지에서 개수를 사용할 수 있도록...?

hongbiii replied:
제 생각에는 개수만 내려주는 API가 추가되는 게 좋을 것 같아요~!

> MainPage에서 목록을 불러와서 context로 하위 페이지에서 개수를 사용할 수 있도록...?

이렇게도 가능하긴 하지만 아직 레이블 페이지나 마일스톤 페이지에 접근하지 않았는데도 목록을 다 불러오는 건 불필요하다는 생각이 들어서요!

hongbiii commented on fe/src/components/Table/MilestonesTable/MilestonesTableItem.tsx:
```diff
@@ -0,0 +1,210 @@
+  const onMilestoneDelete = async () => {
+    try {
+      const res = await deleteMilestone(milestoneId);
+      if (res.status === 204) {
```
status code가 변수로 관리되면 더 좋겠네요 !

hongbiii commented on fe/src/pages/MainPage/IssueDetailPage.tsx:
```diff
@@ -1,68 +1,61 @@
-  );
+  const { data: issueDetails, setData: setIssueDetails } =
+    useFetch<IssueDetails>(
+      useCallback(() => getIssueDetails(issueNumber), [issueNumber])
```
여기 useCallback 은 왜 사용하신 건가요?

youzysu replied:
fetchFn에 arguments가 필요한 경우 익명함수로 한번 감싸야 했는데, 그럼 컴포넌트 렌더링될 때마다 해당 함수가 새로 정의되고, 참조값이 달라져서 useFetch의 useEffect가 무한히 실행되는 문제가 있었습니다!

hongbiii commented on fe/src/pages/MainPage/LabelPage.tsx:
```diff
@@ -1,3 +1,95 @@
+  const navigate = useNavigate();
+
+  const { data: labelsList } = useFetch(getLabels);
+  const { data: milestonesList } = useFetch(getMilestones);
```
(버그) getMilestones state의 디폴트값이 open 이어서 label 페이지에서는 탭에 열린 마일스톤 개수만 표시되네요
디폴트는 전체 검색으로 하고 필터링이 필요할 때만 쿼리스트링을 붙이는 게 좋.......나? 생각해보다가
label 페이지에서는 마일스톤 데이터가 다 필요하지는 않으니
카운트만 내려주는 API가 따로 있으면 좋겠다는 생각이 들었어요!

다음 프로젝트 하실 때 참고해주심 되겠습니다. 😉

youzysu replied:
이 부분은 인지하고 있었는데, 현재로서는 마일스톤이 open/close API만 있는 상황이라서 전체 조회라도 추가로 요청하는게 좋겠네요!

hongbiii commented on fe/src/components/common/TextArea/TextAreaContainer.tsx:
```diff
@@ -0,0 +1,37 @@
+} & React.TextareaHTMLAttributes<HTMLTextAreaElement>) {
+  return (
+    <StyledTextAreaContainer>
+      <TextAreaWrapper {...{ value, ...props }} />
+      <FileUploadArea {...{ appendContent }} />
```
객체 한 겹 벗겨도 될 것 같은데 요렇게 해도 잘 동작하나요?

```suggestion
      <TextAreaWrapper { value, ...props } />
      <FileUploadArea { appendContent } />
```

youzysu replied:
엇 아니요..!
아래 방식 중 하나로 작성해줘야 하네요.
```
<FileUploadArea appendContent={appendContent} />
<FileUploadArea {...{ appendContent }} />
```

hongbiii replied:
아앙 appendContent가 객체인줄 알았네요 굳굳!

hongbiii commented on fe/src/components/Table/MilestonesTable/MilestonesTable.tsx:
```diff
@@ -0,0 +1,36 @@
+
+  return (
+    <Table>
+      <MilestonesTableHeader
+        {...{ currentTabName, handleTabClick, numOpen, numClosed }}
+      />
+      <MilestonesTableBody milestoneList={currentList} />
```
여기도 객체 한 겹 벗기기!

[궁금] props 넘길 때 객체 하나 만들어서 통째로 넘기는 것과 prop 하나씩 넘기는 기준이 있나용? 섞여있어서 여쭤봅니다!

youzysu replied:
props 이름이 동일할 때는 주로 객체로 넘기는 편인데, prop이 하나밖에 없을 때는 직접 넘기기도 합니다. 근데 상단에서 통째로 넘기도록 작성하다보면 하나일 때도 객체로 만들게 되네요 😇 통일성있게 짜는게 좋겠죠

hongbiii commented on fe/src/components/Comment.tsx:
```diff
@@ -1,48 +1,159 @@
+  updateContent: (newContent: string, commendId?: number) => void;
 }) {
+  const [isEditing, setIsEditing] = useState<boolean>(false);
+  const [prevContent, setPrevContent] = useState<string>(content);
```
이 상태는 필요없을 것 같은데 지워도 괜찮은가용?
그냥 props로 받은 content랑 비교하면 되는 것 아닌가 해서요!

youzysu replied:
요청한 데이터가 오기 전 부모에서 content 초기값을 "" 빈 string으로 설정한 것이 Comment props로 전달돼서 이를 감지하기 위해 prevContent 상태가 필요했어요. (IssueDetailHeader 컴포넌트에서도 비슷한 로직이 있습니다.)

이렇게 props가 변경됐을 때 어떻게 해야 하나 찾아보다가 공식 문서를 참고했습니다.
> [prop 변경에 대한 응답으로 특정 state 일부를 조정하려면 렌더링 중에 설정하세요.](https://react-ko.dev/learn/you-might-not-need-an-effect#adjusting-some-state-when-a-prop-changes)

카카모토비가 빈 초기값으로 설정해두고 UI를 그리는게 UX적으로 더 좋을 것 같다고 의견을 주셨고, 저도 동의해서 그렇게 진행했는데, 작업하면서도 빈 초기값을 줬을 때 상태 관리가 너무 까다롭기도 했고, 로직도 끔찍하다는(ㅋㅋㅋ) 마스터의 피드백을 받았어요.

그래서 제 생각에는 지금처럼 빈값으로 UI를 만들기보다 데이터가 왔을 때만(data가 null이 아닐 때만) 해당 컴포넌트를 렌더링하고, 그전에 (isLoading이 true일 때) 유저에게 보여줄 컴포넌트(ex. 스켈레톤)를 **분리해서 별도로** 만들어서 사용하는게 낫다는 생각이 드는데, Autumn은 어떻게 처리하시는 편인가요?

사실 별도로 만들어서 관리하는게 좋겠죠..? 스켈레톤을 만들자니 너무 데이터가 금방 와서 바로 렌더링되네요..😇

hongbiii replied:
넹넹! 보통 loading, success, error 3가지 경우로 나눠서 UI를 그려요.

> 스켈레톤을 만들자니 너무 데이터가 금방 와서 바로 렌더링되네요..😇

데이터가 너무 금방 오면 오히려 스켈레톤이 깜빡거리는 것처럼 보이기도 하긴 하죵 ㅠㅋㅋㅋ
https://tech.kakaopay.com/post/skeleton-ui-idea/
관련해서 위 아티클을 재밌게 읽었던 기억이 나네요!

hongbiii commented on fe/src/components/Comment.tsx:
```diff
@@ -1,48 +1,159 @@
+    try {
+      const { status } = commentId
+        ? await putIssueComment(issueNumber, commentId, body)
+        : await putIssueContent(issueNumber, body);
```
줄바꿈이 절묘하게 되어 있어서 commentId를 디스트럭쳐링 하는 건줄 알았어요..!
조금 중복이 생기더라도 이때는 함수 자체를 분리하면 어떨까 싶습니당

```typescript
const onEditContent = () => {
  // ...
}

const onEditComment = () => {
  // ...
}

```

hongbiii commented on fe/src/components/Issues/IssueDetail/IssueCommentContainer.tsx:
```diff
@@ -0,0 +1,158 @@
+        return comment;
+      });
+    });
+  };
```
수정 성공 시 리액트 상태를 기반으로 리렌더링을 해주고 있는데

성공 시 -> get 요청 다시 한다.
실패 시 -> 적절한 에러 ui를 노출한다.

의 플로우로 가면 어떨까요?
가능하면 서버 데이터와 싱크를 유지하는 게 좋다고 생각해서요!

hongbiii reviewed (approved):
조이, 카카모토비 안녕하세요~~ 이번 주도 고생 많으셨어요!!

전체적으로 요청 성공 시 리렌더링하는 부분에서 이전 상태값을 기반으로 상태를 업데이트해서 리렌더링하는 방식이 사용되고 있네용
코멘트에도 언급했지만 로컬 state를 업데이트 하는 것보다 GET 요청을 다시 보내는 방식이 더 좋다고 생각하는데요!
이유는 서버 데이터와 싱크를 맞추는 게 좋다고 생각해서입니당 😃
본문에 작성해주신 것처럼 refetch를 사용하면 해결이 될 것 같아요~!

에러 처리 try catch로 꼼꼼하게 잘 해주셨네요 !
useFetch 잘 만들어주셨으니 이 훅에 요청 성공, 실패 콜백함수를 넘겨주는 방법으로 리팩토링 해보는 것도 좋을 것 같습니당

youzysu commented on fe/src/components/Issues/IssueDetail/IssueDetailHeader.tsx:
```diff
@@ -0,0 +1,216 @@
+    };
+
+  const [isEditTitle, setIsEditTitle] = useState(false);
+  const [newTitle, setNewTitle] = useState(title);
+  const [prevTitle, setPrevTitle] = useState(title);
+
+  if (title !== prevTitle) {
+    setPrevTitle(title);
+    setNewTitle(title);
+  }
```
참고: 해당 부분

youzysu reviewed (commented):
이번주도 유익한 리뷰 감사합니다 Autumn! 🥰
일정 때문에 코드 퀄리티가 저하되고 있는데 다음 주에 리팩토링해보려고 합니다 흑흑
데이터 요청 전 렌더링 방식에 대해서 궁금한 부분만 추가로 질문 남겼는데([링크](https://github.com/codesquad-members-2023/issue-tracker-max/pull/56/files/45f1ee1606aba6aad2fff543463a5ecf0f458918#r1292692674)), 추가로 답변해주실 수 있을까요~? 참고해서 진행하려고 합니다! 감사합니다!!!

youzysu committed:
[FE] 3주차 리뷰 반영 리팩토링 (#145)

* #144 refactor: useFetch 개선 & 이슈 상세 페이지 적용

* #144 refactor: 마일스톤 페이지 상태 관리 방식 적용

* #144 refactor: 레이블 페이지 상태 관리 방식 적용

* fix: 마일스톤 페이지 NaN 수정, 스타일 개선

- ProgressBar percent NAN 수정: 열린 이슈, 닫힌 이슈 개수 optional로 수정
- variant에 따라 버튼 문구 반영

* #144 rename: useFetch 내부 함수명 수정

- updateData -> reFetch
